### DISCUSSION TITLE
ErrorOutOfBounds for tracer field sampling

### DISCUSSION THREAD

**Discussion by brogalla** [Parcels-code/Parcels#1268] [Q&A]
Hi Parcels Community,

I'm running into an ErrorOutOfBounds problem when sampling a tracer field based on particle tracks and was hoping that one of you might have an idea what the issue is. When I release particles at a particular location + time without field sampling, the particles run fine. However, running the same particles with my tracer sampling kernel, I get ErrorOutOfBounds for some, but not all particles. I've checked that the latitude and longitudes in the fieldsets for the tracer and velocity fields cover the sampled locations.

For my code, I tried to follow a method similar to this [example notebook](https://nbviewer.org/github/OceanParcels/parcels/blob/master/parcels/examples/tutorial_sampling.ipynb?flush_cache=true). You can see my notebook with the code that produces the problem [here](https://github.com/UBC-MOAD/analysis-brogalla/blob/main/particle_tracking/experimentation/run-parcels-tracer-field.ipynb). I'm running with parcels version 2.4.0.

Thanks in advance!

**Comment by erikvansebille** [comment]
Thanks for reporting, @brogalla. I think the issue is with the depth of the particles. When you don't specify a `depth` in your ParticleSet, Parcels takes the shallowest depth of the U-Fields. But the Tracer field only starts deeper, as you see in your last cell:
```python
print(np.amin(fieldset.U.depth), np.amax(fieldset.U.depth))
print(np.amin(fieldset.Pb.depth), np.amax(fieldset.Pb.depth))
0.0 5500.0015
0.49402538 5727.9165
```
So the solution could be to create your particles with e.g. `depth=0.5`?

**Reply by brogalla** [reply]
Thanks for the quick response, Erik! Good point about the release depth, I think that fixed one underlying issue. If I release the particles at depth=1.0, the code is happy with 1 more of the particles, but the other two are still out of bounds. 

Looking at it more, I suspect the issue is related to the tracer field being on a sub-domain of the full grid. When I use the W velocity (on full grid) as a tracer and sample it, I don't get an error. Maybe I'm misunderstanding how to correctly specify the sub-grid indices (I also get the error if I don't specify sub grid indices)? 
- U, V, and W variables are on a full domain with dimensions z,y,x (50,2400,1632) 
- tracer Pb is on a sub domain of the full grid with dimensions z,y,x (50,1040,640) starting at the lower left corner of indices (1139, 159) of the full domain

I currently specify the fieldset indices to take into account the subset grid as:
```
indices_subset = {'U':  {'lon': range(159, 799), 'lat': range(1139, 2179), 'depth': range(0,50)},
                  'V':  {'lon': range(159, 799), 'lat': range(1139, 2179), 'depth': range(0,50)},
                  'W':  {'lon': range(159, 799), 'lat': range(1139, 2179), 'depth': range(0,50)},
                  'Pb': {'lon': range(0,640)   , 'lat': range(0,1040)    , 'depth': range(0,50)}}
```

**Reply by erikvansebille** [reply]
Well, in this case you would not need the indices for `Pb`, so you could drop them from the dictionary

Strange that you also get the error when you don't specify the indices at all; that was going to be my next suggestion. What do you get as lon/lat/depth ranges when you print them, in that case? Your data is not over the dateline, right?

**Reply by brogalla** [reply]
The data doesn't go over the dateline. Without indices I have:
```Python
print(np.amin(fieldset.U.lon), np.amax(fieldset.U.lon))
print(np.amin(fieldset.Pb.lon), np.amax(fieldset.Pb.lon))
print(np.amin(fieldset.U.lat), np.amax(fieldset.U.lat))
print(np.amin(fieldset.Pb.lat), np.amax(fieldset.Pb.lat))
print(np.amin(fieldset.U.depth), np.amax(fieldset.U.depth))
print(np.amin(fieldset.Pb.depth), np.amax(fieldset.Pb.depth))
-179.9996 179.99995
-176.72366 -32.57282
-20.07611 89.95883
55.4229 85.85598
0.0 5500.0015
0.49402538 5727.9165
```
The error is raised for the particles for lon, lat, depth: (-130, 75, 1) and (-145, 74, 1); but (-70, 65, 1) and (-55, 65, 1) are fine.

**Reply by erikvansebille** [reply]
Hmm, that is strange. I don't know what is going on then. Could you send me some of your files so that I can run some tests myself?

**Reply by brogalla** [reply]
Definitely, thanks! I've sent a link to a few example files to your email address; let me know if you'd prefer me to send it some other way.

**Reply by erikvansebille** [reply]
Hi @brogalla, I had a look at your Issue this afternoon, but I think this is a more fundamental problem for Parcels: it can't easily handle non-monotonous grids....

Your Grid is extremely curvilinear, which you can see if you issue 
```python
fieldset.Pb.show()
```
![image](https://user-images.githubusercontent.com/14315062/203412977-4f0ad558-d862-4641-a575-954bb0299817.png)

Now,  try to fill your entire domain with Particles, by creating a (larger) regular grid of these and seeing which ones 'stick' when you delete all the ones out of bounds:
```python
x = np.arange(-180, 180, 1)
y = np.arange(55, 89, 1)
plon, plat = np.meshgrid(x, y)
pset = ParticleSet(fieldset=fieldset, pclass=JITParticle, lon=plon, lat=plat)

pset.execute(AdvectionRK4_3D, runtime=0, dt=0, recovery={ErrorCode.ErrorOutOfBounds: DeleteParticle})

plt.plot(pset.lon, pset.lat, 'k.')
```

You see that all the ones west of approx 120W get deleted... 
![image](https://user-images.githubusercontent.com/14315062/203413850-4dcac810-94eb-4ad6-9598-52a558a4d4df.png)

So my hunch is that this is not a problem with the Pb tracer field, but rather a problem that your grid is so extremely curved. The way that Parcels tries to find in which grid cell a particle is located (based on its lat/lon) just doesn't work for this Grid. 

It will be a lot of under-the-hood development to fix this. So I'm very sorry, but I don't think that Parcels will work on your particular Grid...

**Reply by erikvansebille** [reply]
Actually, what I just wrote about difficult to fix this may not be entirely true. It occurred to me that you could _help_ parcels by trying to find the indices of the grid cells yourself. Once these indices are correctly found, Parcels does a local search to update a grid cell when a particle moves out, but this is much simpler and more robust. So only the initial search is the problem

If you add the code below after you create the `ParticleSet` (but before the execute), things may work?

```python
for i in range(len(pset)):
    d = np.hypot((pset.lon[i] - fieldset.U.grid.lon), (pset.lat[i] - fieldset.U.grid.lat))
    ind = np.unravel_index(np.argmin(d, axis=None), d.shape)
    pset.xi[i, :] = ind[1]
    pset.yi[i, :] = ind[0]
```
See also the image below, which now shows _all_ particles in your curved domain
![image](https://user-images.githubusercontent.com/14315062/203422943-3653389c-65f4-4309-95c8-df530a9f3746.png)

And it also works with sampling now:
```python
import matplotlib.colors as cols
plt.scatter(pset.lon,pset.lat, c=pset.dissolpb, s=5, norm=cols.LogNorm())
```
![image](https://user-images.githubusercontent.com/14315062/203423997-27d866a0-cb06-4192-b338-d47308d627e6.png)

**Reply by brogalla** [reply]
Thanks for looking into this more, Erik! Much appreciated. I did some more digging and it's odd... Although the curvature of the grid certainly could be part of the problem, I suspect it's not the main one. The subdomain and full domain have the exact same underlying grid (one is a subset of the other), so their curvature should be identical. However, the issue arises only with sampling the tracer field, so when I run particles with the velocities from the full domain and don't sample the tracer field, everything runs fine and OP doesn't have trouble finding the correct cell locations. You can see my debugging in [this notebook](https://github.com/UBC-MOAD/analysis-brogalla/blob/main/particle_tracking/experimentation/tracer-field-debugging.ipynb). 

I tried your nice suggestion of hinting at the coordinates for the start locations (I'll call it "nudging"). It definitely helps it start! When I run for a few timesteps, some of the parcels remain while the rest are removed because the correct cell location is not found within 100000 iterations: 
![image](https://user-images.githubusercontent.com/13894889/204066248-66bd7be9-1f6b-43da-b114-ef6be348f91b.png)

When I run without indices specified in the Fieldset and without sampling the tracer, I don't need the nudging in order to run the particles in the entire region, while it was necessary when the indices were specified. For the next part, I've removed the indices in the Fieldset. If I run particles for multiple timesteps without sampling the tracer (so just with velocities from the full domain) and without nudging, it runs fine:
![image](https://user-images.githubusercontent.com/13894889/204066259-7f29ad77-bbdb-47fa-8981-7d4d4aba6cba.png)

For the identical setup, if I add the tracer sampling kernel the subdomain is selected but excludes an area from the western end of the subdomain and it can't find a bunch of the particle locations (top figure), while if I nudge the start location the whole subdomain gets shifted which is odd (bottom figure):
![image](https://user-images.githubusercontent.com/13894889/204066283-4b192b59-eb35-4ce7-9981-c0cb3d9dfcc8.png)
![image](https://user-images.githubusercontent.com/13894889/204066287-c6bf6faf-4d05-4520-8fa0-451eed92663c.png)

So... I'm adding more confusion rather than reducing it :sweat_smile:. I'll keep playing around with it; let me know if you have any further suggestions, but I also understand if this is just an odd use case.

**Reply by erikvansebille** [reply]
Hi @brogalla, I had another lok into your issue, and I think I may have found the problem. Your OutOfBounds kernel kills the particles also if they go through the ocean surface. This problem was also recently discussed in e.g. https://github.com/OceanParcels/parcels/discussions/1266, with solution here: https://github.com/OceanParcels/parcels/discussions/1266#discussioncomment-4177129. Not sure why this is, but it may be related to the slight differences in the uppermost depth level of your `U` and `Pb` fields. Can you see if the solution above helps?

(by the way, I also found that the code for 'nudging' the starting locations was not perfect; as it only used the U and V grids; not the Pb grids. Especially when you don't use indexing, these can be very different. The proper code is below, where the correct index is found for each of the grids in the fieldset)

```python
for i in range(len(pset)):
    for gi, g in enumerate(fieldset.gridset.grids):
        d = np.hypot((pset.lon[i] - g.lon), (pset.lat[i] - g.lat))
        ind = np.unravel_index(np.argmin(d, axis=None), d.shape)
        pset.xi[i, gi] = ind[1]
        pset.yi[i, gi] = ind[0]
```

**Reply by brogalla** [reply]
Woohoo! The solution from the through surface problem worked (and should be fine for my purposes) when combined with the corrected nudging! Thank you so much for the help, I really appreciate it and am happy to be able to use OP to look at tracer fields in another way.
